sources/licenses/Licenses.yml

```diff
@@ -45,12 +48,46 @@ swap:
   CC[ -]?BY[ -]?SA[ -]?4(?:\.0)?: CC BY-SA 4.0
   CC[ -]?[O0]: CC0
   CC[ -]?0[ -]?1(?:\.0)?: CC0 1.0
-  # GNU Licences
+  # GNU licences
   AGPL: AGPL
-  AGPL[ -]?v?3(?:\.0)?: AGPLv3
+  AGPL[ -]?v?3(?:\.0)?.*(?!only|or-later): AGPLv3
   GPL(?!\.org): GPL
-  GPL[ -]?v?2(?:\.0)?: GPLv2
-  GPL[ -]?v?3(?!\+): GPLv3
-  GPL[ -]?v?3\+: GPLv3+
+  GPL[ -]?v?2(?:\.0)?.*(?!only|or-later): GPLv2
+  GPL[ -]?v?3(?!\.0)?.*(?!only|or-later): GPLv3
```

Comment:
I strongly disagree on the use of ``.*``

> The .* is a huge problem
> 
> It would match crazy things
> 
> AGPLv3I-like_chicken

Why should we care about that? I may not be understanding the issue here, since I would in fact prefer the rule to correct ``AGPLv3I-like_chicken`` to ``AGPLv3``, rather than avoid it.

> Here is an example of false-negative allowed with the regexp I suggest in this post with GPL-2
> 
> GPL v2-only GPLv2-only GPL v2.0-only GPLv2.0-only GPL v2.0-only GPL-v2.0-only GPL-v2-only GPL v2-or-later GPLv2-or-later GPL v2.0-or-later GPLv2.0-or-later GPL-v2.0-or-later GPL-v2-or-later
>

Allowing or not fixing those phrases shouldn't be acceptable. The rule as I suggested works with the simple test below:

```
## Good
GPLv2
GPLv2+
GPL-2.0-only
GPL-2.0-or-later

## Bad
GPL2
GPL-v2-or-later
GPL-2
GPL-2-only
```

Frankly, I believe that to be a better compromise. I don't think we ought to consider and take into account anything wildly non-conforming like ``GPL-v2-aka-the-license-I-like``. Those oddities are fairly rare, and I'd rather have the rule correct them to the closest approximation, in the example's case to GPLv2.

Note: I've fixed the odd difference between the GPLv2 and the GPLv3 corrections. I'll also look into the missing slashes for the later section of the rule.

Reply:
OK. I'm fine with what you suggest.

But then I would expect all these examples to be added in good.txt with a mention they are allowed for how, but could be changed later

Reply:
About the .* issue my problem is also this

`GPLv2 or Apache 2.0, jargonLint is great, but kung of panda 2-or-later could be better.`

I expect the .* regexp to match all this

``GPLv2 or Apache 2.0, jargonLint is great, but kung of panda 2-or-later`

It's worth addind such example in good/bad.txt

Reply:
I think I've fixed that by changing ``.*`` to ``[^\s]*``, so that it includes everything but spaces. The new expression becomes ``GPL[ -]?v?2(?:\.0)?[^\s]*(?!only|later)`` and doesn't match entire sentences. Do you notice any problems with this one?

Reply:
> But then I would expect all these examples to be added in good.txt with a mention they are allowed for how, but could be changed later

What is allowed for now? Do you mean something, other than the weird sentence match, is allowed when it shouldn't be? ``GPL-2.0-or-later`` should be allowed and ``GPL-v2-or-later`` should be disallowed, both of which the rule achieves. Is there something I'm missing?

Reply:
Regexp are nightmare to maintain.

Adding examples to what is supported now is always a good thing for when the regexp will be updated.

Your current regexp works but you are accepting more than the spdx identifiers you wanted to allow

So anything invalid followed by a -only or later is accepted by the current regexp

GPL-v2.0-only
GPLv2.0-or-later

I might be wrong, it's uneasy to check with a phone

I'm just asking you to add the use cases to bad and good and see what's happening now, and guarantee what will happen later if the regexp is updated

Reply:
What was tge problem with the regexp I provided earlier?

GPL[ -]?v?2(?:\.0)?(?!\-only|\-or\-later): GPLv2

This one is limited to a very small subset of characters

Reply:
I'll leave the screenshots here to illustrate the difference. 

Your rule suggestion corrects partially and the correction creates another problem. The worse issue, and this was the case with my initial suggestion as well, is that correct phrases like ``GPL-2.0-only`` are not excluded.

![image](https://github.com/user-attachments/assets/af9d4ef7-e9f3-4fc7-bb47-dd9e8c6b0c9d)

The beauty of the latest rule suggestion, ``GPL[ -]?v?2(?:\.0)?[^\s]*(?!only|later)``, is that the exceptions are properly excluded while phrases like ``GPL-v2.0-only`` are still fixed.

![image](https://github.com/user-attachments/assets/f538719f-2a30-46f7-969d-95cac4f45a92)

Link to test the latest rule: https://studio.vale.sh/s/d807839ad15d7ec5a69afe831fd0b460

Reply:
Thanks a lot it's clearer with a link to the studio vale

Reply:
Resolving this issue then.